cpp/src/arrow/json/reader_test.cc

```diff
@@ -274,5 +275,35 @@ TEST(ReaderTest, ListArrayWithFewValues) {
   AssertTablesEqual(*actual_table, *expected_table);
 }
 
+TEST_P(ReaderTest, UnquotedDecimal) {
+  parse_options_.unexpected_field_behavior = UnexpectedFieldBehavior::InferType;
+  auto schema =
+      ::arrow::schema({field("price", decimal(9, 2)), field("cost", decimal(9, 3))});
+  parse_options_.explicit_schema = schema;
+  auto src = unquoted_decimal_src();
+  SetUpReader(src);
+  ASSERT_OK_AND_ASSIGN(table_, reader_->Read());
+
+  auto expected_table = Table::Make(
+      schema, {ArrayFromJSON(schema->field(0)->type(), R"(["30.04", "1.23"])"),
+               ArrayFromJSON(schema->field(1)->type(), R"(["30.001", "1.229"])")});
```

Comment:
Nit: these could use a simpler helper
```suggestion
  auto expected_table = TableFromJSON(schema, {R"([
    { "price": 30.04, "cost":30.001 },
    { "price": 1.23, "cost":1.229 }
  ])"});
```

Reply:
Changed to TableFromJSON